### ISSUE TITLE
MouseEvent.fromElement/toElement

### ISSUE THREAD

**Issue by cvrebert** [w3c/uievents#84]
Consider either spec'ing or trying to kill `MouseEvent.fromElement` & `MouseEvent.toElement`. Here's the current (in)compatibility situation:

|  Events                                          | Chrome+Safari               | Edge                        | Firefox |
| --- | --- | --- | --- |
| `mouseenter`/<br>`mouseleave`/<br>`mouseover`/<br>`mouseout` | `fromElement` & `toElement` | `fromElement` & `toElement` | Neither |
| `click`                                          | `toElement` only            | Neither                     | Neither |

Docs:
* http://www.quirksmode.org/js/events_mouse.html#relatedtarget
* https://msdn.microsoft.com/en-us/library/ms533773(v=vs.85).aspx
* https://msdn.microsoft.com/en-us/library/ms534684(v=vs.85).aspx

Testcase: http://jsbin.com/vivigi/edit?html,js,output

**Comment by cvrebert** [comment]
Chrome usage metrics:
* `fromElement`: https://www.chromestatus.com/metrics/feature/timeline/popularity/506
* `toElement`: https://www.chromestatus.com/metrics/feature/timeline/popularity/507

**Comment by cvrebert** [comment]
Proposed tests: https://github.com/w3c/web-platform-tests/pull/4408
(Doesn't cover the `click` case)

**Comment by cvrebert** [comment]
Given the >20% usage numbers from chromestatus, it seems **very** unlikely that these events will be removable.

**Comment by karlcow** [comment]
@cvrebert looking at https://www.chromestatus.com/metrics/feature/timeline/popularity/507 about `toElement` which has 40% usage!!! I wonder why there isn't not that many webcompat reports for Firefox. It sounds like there's a hidden story behind this.

The only webcompat report I can track down for now is
https://bugzilla.mozilla.org/show_bug.cgi?id=1205244

where relatedTarget has to be used instead. 

My feeling is that if so many pages are using toElement, it must be a case of  toElement || relatedTarget thing or something similar?

It would be to know which pages we see this toElement.

@annevk @foolip

**Comment by foolip** [comment]
The usage will be dominated by code that copies events by enumerating all properties. The prevalence of this type of code makes use counters on many event properties totally misleading.

**Comment by hsivonen** [comment]
[Another Gecko bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1539619) (Assumes that these exist if `window.event` exists.)

**Comment by foolip** [comment]
@miketaylr these two attributes came up in your analysis of APIs missing from Firefox but with high usage in Chrome, and you said it didn't seem like a compat problem because usage was driven by jQuery. Seems like maybe it is a problem after all?

**Comment by foolip** [comment]
Here's the implementation in WebKit and Chromium:
https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/dom/MouseEvent.cpp?rev=235865#L211
https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/events/mouse_event.cc?l=367&rcl=d95c281fff5f43cc6fa7e728bb62b028ad3c9cc7

They haven't diverged in their observable behavior, and writing down + testing these steps should be straightforward. @garykac is this something you could take a look at, or would you accept PRs?

**Comment by miketaylr** [comment]
> you said it didn't seem like a compat problem because usage was driven by jQuery. Seems like maybe it is a problem after all?

Yeah, we didn't have any issues, but it looks like shipping `window.event` opened a possible can of worms. I would guess the bug linked by @hsivonen won't be a unique instance.

**Comment by garykac** [comment]
I can certainly add this to the spec (or accept PRs, whichever).

Confirming that the consensus to:
* Keep `relatedTarget` spec unchanged
* Add `toElement`/`fromElement` spec based on Chrome/Webkit impl

AFAICT, only Mozilla will be impacted by adding that to the spec. If they are OK with that, then I have no problems with doing this.

**Comment by karlcow** [comment]
if Mozilla engineers were to implement `fromElement`/`toElement`, it would require to understand the consequences of doing so. :) More or less breakages.

**Comment by foolip** [comment]
@garykac is there a difference between the spec and implementations for `relatedTargets`, or why do you ask if that should remain unchanged? If there is a difference, then it's probably still fine to define `toElement` and `fromElement` in terms of `relatedTarget` and try to align `relatedTarget` separately.

**Comment by foolip** [comment]
Alright! :) Also in the interest of being explicit, do you plan to write a PR?

**Comment by garykac** [comment]
Yes!

**Comment by foolip** [comment]
Sweet, thank you!

**Comment by defaude** [comment]
Hey everyone! I came across this issue when I was researching a bug when researching the "missing" `toElement` property in Firefox when looking at a `mouseleave` event. :)

@garykac did you find the time to prepare the PR already? Or is there something I can help with?

**Comment by appsforartists** [comment]
Stumbled upon this trying to figure out how to implement dragging a file into a web page.

The problem is that the drag-and-drop API is very finicky.  There is a firehose of events that fire during a drag, a mix of `dragenter`, `dragover`, and `dragleave`s.  Even though you have to call `event.preventDefault()` inside a `dragover` to claim responsibility for a drag, the events are all dispatched from children.  The naive implementation - setting a style inside `dragover` and removing it in `dragleave` - causes flickering, where the style is present in odd numbered frames and removed in even numbered frames, because the children are being entered/left even though the drag is still hovering over the drop target.

The current best practice is to count the number of `dragenter` and `dragleave` events, using the balance to determine whether or not the drop target is being hovered over.  It's a hack to paper over an unfortunate API.

`fromElement` provides an opportunity to detect whether the `dragleave` event is actually escaping the drop target:

```typescript
onDragOver(event) {
  // this e.pD() is important to register for the other events
  event.preventDefault();
  setInDrag(true);
},
onDragLeave(event) {
  let fromAncestor = event.fromElement;
  while (fromAncestor = fromAncestor.parentNode) {
    if (fromAncestor === dropTarget) {
      return;
    }
  }
  setInDrag(false);
},
onDrop(event) {
  event.preventDefault();
  setInDrag(false);

  const fileList = event.dataTransfer.files;
  //…
},
```
Standardizing `fromElement` would improve the drag-and-drop API.
